File: android/chapbook/src/main/kotlin/com/ophymx/chapbook/Session.kt
package com.ophymx.chapbook

import android.graphics.Bitmap
import android.os.ParcelFileDescriptor

/** Where the reader is. The pair, never the page alone — see `docs/SHELLS.md`. */
data class Position(val spine: Int, val page: Int)

/** What kind of book a session opened. */
enum class BookKind { EPUB, COMIC, PDF }

/** The engine's colour themes, in cycle order. */
enum class Theme { LIGHT, SEPIA, DARK }

/**
 * The scalar reading settings. The font family deliberately travels on
 * its own calls — see [Session.setSettings].
 */
data class ReadingSettings(
    val baseFontPx: Float,
    val lineHeight: Float,
    val justify: Boolean,
    val publisherStyles: Boolean,
    val theme: Theme,
)

/**
 * Something the session wants a shell to know, drained from
 * [Session.drainEvents]. Deliberately not about drawing — what to repaint
 * is [Session.pollLoaded]'s answer.
 */
sealed class SessionEvent {
    /** A background unit finished decoding, prefetches included. */
    data class UnitLoaded(val spine: Int) : SessionEvent()

    /** A background unit failed and will not be retried. [message] is for a person. */
    data class UnitFailed(val spine: Int, val message: String) : SessionEvent()

    /** The reader is somewhere else — moves the shell did not make included. */
    data class PositionChanged(val spine: Int, val page: Int) : SessionEvent()

    /** The reader reached the last page of the last unit. */
    data object BookFinished : SessionEvent()
}

/** The device-pixel size a bitmap must be before [Session.renderInto] will draw. */
data class RenderSize(val width: Int, val height: Int)

/**
 * One visual line of the current page, with its geometry and locator
 * range — the material an `AccessibilityNodeInfo` tree is built from.
 * The rect is page space (logical units, page top-left); the locator
 * range is `[start, end)` in the same offsets positions and annotations
 * use. The text's length is not the locator span's width.
 */
data class TextRun(
    val text: String,
    val rect: android.graphics.RectF,
    val locatorStart: Int,
    val locatorEnd: Int,
)

/**
 * One word: where it sits in [Session.speakableText] (char offsets) and
 * in locator space. `onRangeStart` from a TTS utterance reports offsets
 * into the string; the locator range is how that progress becomes a
 * highlight via [Session.rangeRects].
 */
data class WordSpan(
    val textStart: Int,
    val textEnd: Int,
    val locatorStart: Int,
    val locatorEnd: Int,
)

/**
 * What the engine did with an action, and what you owe the platform back.
 *
 * Two questions, not one, and neither implies the other. [needsRedraw] says
 * whether to `invalidate()`. [consumed] says what to return from
 * `onKeyDown`/`onKeyUp` — and getting *that* wrong is expensive here,
 * because the default key map binds the volume keys to page turns: a
 * reader that answers "nothing changed" on the last page hands the press
 * back and Android draws its volume slider over the book.
 */
enum class ActionOutcome {
    /** Applied, and something moved. Repaint, and consume the event. */
    Changed,

    /** Applied, nothing moved — last page, font at its stop. Still yours. */
    Unchanged,

    /**
     * Not the engine's: `toggle-menu` always, and `back` with an empty
     * trail. Let the event through, which is how the system Back leaves
     * the reader without this side tracking the history to know when.
     */
    Unhandled;

    val needsRedraw: Boolean get() = this == Changed
    val consumed: Boolean get() = this != Unhandled
}

/**
 * One open book.
 *
 * Held by the app for as long as it is reading, and [close]d exactly once.
 * The underlying session is `Send` but not `Sync`: it may move between
 * threads, and must never be touched from two at the same time. This class
 * does not enforce that — the C ABI's header will say it, and a real
 * binding should.
 */
class Session private constructor(private var handle: Long) : AutoCloseable {

    companion object {
        /**
         * Send the engine's diagnostics to logcat under the tag `chapbook`.
         *
         * Call once, before opening anything. The engine is silent until a
         * host installs a backend, and on Android silent used to mean the
         * failures only a device hits were the ones nobody could see.
         */
        fun initLogging(verbose: Boolean = false) = Native.initLogging(verbose)

        /**
         * Opens a book from a filesystem path.
         *
         * `libraryDir` should be `context.filesDir` — it is where positions,
         * annotations and settings live, and it is an argument because a
         * sandboxed app's answer is not reachable any other way.
         *
         * Returns null if the book will not open; logcat says why.
         */
        fun open(path: String, libraryDir: String): Session? {
            val handle = Native.open(path, libraryDir)
            return if (handle == 0L) null else Session(handle)
        }

        /**
         * Opens a book from a `content://` URI's file descriptor — what the
         * storage access framework actually hands an app, with no path and
         * usually no extension. The format comes from the bytes.
         *
         * Takes ownership of [pfd]: it is detached here and closed by the
         * session, so the caller must not close it or use it again.
         *
         * A book opened this way reaches the library by content — hashed on
         * open, recorded under the same edition fingerprint a path import
         * gets — so its position and annotations persist. Reopening the
         * *file* next launch is the app's job: take a persistable URI
         * grant, re-resolve it, and hand the descriptor back here.
         */
        fun openFd(pfd: ParcelFileDescriptor, libraryDir: String): Session? {
            val handle = Native.openFd(pfd.detachFd(), libraryDir)
            return if (handle == 0L) null else Session(handle)
        }

        /** Run the conformance harness over a book. It opens its own sessions. */
        fun conformance(path: String, libraryDir: String): String =
            Native.conformance(path, libraryDir)
    }

    /**
     * What the font source produced: faces loaded, and any generic family
     * pointed at a name no loaded face carries.
     *
     * Zero faces means every page paginates blank, taking navigation,
     * search and the table of contents with it — still the single most
     * useful string on this screen.
     */
    val fontReport: String get() = Native.fontReport(handle)

    val title: String get() = Native.title(handle)

    val position: Position
        get() {
            val packed = Native.position(handle)
            return Position((packed ushr 32).toInt(), (packed and 0xffffffffL).toInt())
        }

    /** Null until [setMetrics] has been called. */
    val renderSize: RenderSize?
        get() {
            val packed = Native.renderSize(handle)
            if (packed < 0) return null
            return RenderSize((packed ushr 32).toInt(), (packed and 0xffffffffL).toInt())
        }

    val cacheBytes: Long get() = Native.cacheBytes(handle)
    val cacheBudget: Long get() = Native.cacheBudget(handle)

    fun setMetrics(width: Float, height: Float, margin: Float, scale: Float) =
        Native.setMetrics(handle, width, height, margin, scale)

    /** Returns whether the position moved. Do not derive this from [position]. */
    fun nextPage(): Boolean = Native.nextPage(handle)

    /** Returns whether the position moved. */
    fun prevPage(): Boolean = Native.prevPage(handle)

    /** Next theme. Repaints; the position does not move. */
    fun cycleTheme() = Native.cycleTheme(handle)

    // ---- Background loads ----
    //
    // Image books (CBZ, PDF) decode off the UI thread. Without this pair
    // wired, a comic opens to its placeholder page and stays there.

    /**
     * Install the wake callback, run once per landed load **on the loader
     * thread**. It must only get back to the main thread — `View.post`, a
     * `Handler` — and call [pollLoaded]; touching a view from inside it is
     * the bug every reference shell's waker comment warns about. Null
     * clears it.
     */
    fun setWaker(waker: Runnable?) = Native.setWaker(handle, waker)

    /**
     * Take delivery of anything the loader finished. Returns whether the
     * visible page changed, and therefore whether to repaint — prefetch
     * landings answer false on purpose.
     */
    fun pollLoaded(): Boolean = Native.pollLoaded(handle)

    /** Whether any unit is still loading — for a shell that shows a spinner. */
    val hasPendingLoads: Boolean get() = Native.hasPendingLoads(handle)

    // ---- Session events ----

    /**
     * Everything the session wants a shell to know since the last drain:
     * loads landing and failing, the position moving, the book finishing.
     * Drain after a wake or a draw; the engine coalesces on its side.
     */
    fun drainEvents(): List<SessionEvent> {
        val out = mutableListOf<SessionEvent>()
        while (true) {
            val packed = Native.nextEvent(handle)
            if (packed < 0) break
            val spine = ((packed ushr 28) and 0x0fff_ffff).toInt()
            val page = (packed and 0x0fff_ffff).toInt()
            out.add(
                when ((packed ushr 56).toInt()) {
                    0 -> SessionEvent.UnitLoaded(spine)
                    1 -> SessionEvent.UnitFailed(spine, Native.eventMessage(handle) ?: "")
                    2 -> SessionEvent.PositionChanged(spine, page)
                    else -> SessionEvent.BookFinished
                }
            )
        }
        return out
    }

    // ---- The rest of the reading model ----

    /** How many spine units the book has — the denominator of "ch 2/8". */
    val spineLen: Int get() = Native.spineLen(handle)

    /** How many pages the current unit laid out to; 0 until metrics arrive. */
    val pageCount: Int get() = Native.pageCount(handle)

    /** What kind of book — "ch" or "pg" in a title bar. */
    val kind: BookKind
        get() = when (Native.bookKind(handle)) {
            1 -> BookKind.COMIC
            2 -> BookKind.PDF
            else -> BookKind.EPUB
        }

    // ---- Settings ----

    /** The settings in force, resolved override-over-default. */
    val settings: ReadingSettings?
        get() {
            val values = Native.settings(handle)
            if (values.size < 5) return null
            return ReadingSettings(
                baseFontPx = values[0],
                lineHeight = values[1],
                justify = values[2] != 0f,
                publisherStyles = values[3] != 0f,
                theme = Theme.entries.getOrElse(values[4].toInt()) { Theme.LIGHT },
            )
        }

    /**
     * Replace the scalar settings. The chosen font family is preserved,
     * not cleared — it travels on [setFontFamily], the same split the C
     * ABI keeps and for the same reason. [thisBook] scopes the change to
     * this book instead of the reader default; both persist through the
     * library when the session has one.
     */
    fun setSettings(settings: ReadingSettings, thisBook: Boolean = false) =
        Native.setSettings(
            handle,
            settings.baseFontPx,
            settings.lineHeight,
            settings.justify,
            settings.publisherStyles,
            settings.theme.ordinal,
            thisBook,
        )

    /** The reader's chosen typeface, or null for the publisher's. */
    val fontFamily: String? get() = Native.fontFamily(handle)

    /** Choose a typeface — one of [fontFamilies] — or null to let go. */
    fun setFontFamily(family: String?, thisBook: Boolean = false) =
        Native.setFontFamily(handle, family, thisBook)

    /** Every family the session's font database offers — a picker's list. */
    val fontFamilies: Array<String> get() = Native.fontFamilies(handle)

    /**
     * Save the position and drop everything reconstructible. Call from
     * `onStop`, which is the last callback Android guarantees.
     */
    fun suspend() = Native.suspendSession(handle)

    /** Call from `onTrimMemory`. The current page is rebuilt on the next draw. */
    fun releaseCaches() = Native.releaseCaches(handle)

    /**
     * Draws the current page into [bitmap], which must be `ARGB_8888` and
     * exactly [renderSize]. The engine rasterizes into the bitmap's own
     * pixels; nothing is copied. Returns 0, or a negative code described in
     * `chapbook-jni`.
     */
    fun renderInto(bitmap: Bitmap): Int = Native.renderInto(handle, bitmap)

    /**
     * Which edge this book reads from: `"ltr"` or `"rtl"`.
     *
     * The book declares it — EPUB's `page-progression-direction` — and the
     * tap zones already use it. This is here so a shell can show that it
     * did, because a correctly flipped RTL book and a bug look the same
     * from the outside.
     */
    val readingDirection: String get() = Native.readingDirection(handle)

    /**
     * Reconfigure the tap bands as fractions of the page width. [middle] is
     * an action name, or `""` for a band that does nothing.
     *
     * There is no direction parameter on purpose: that one is the book's.
     */
    fun setTapZones(prevFraction: Float, nextFraction: Float, middle: String) =
        Native.setTapZones(handle, prevFraction, nextFraction, middle)

    /**
     * What a tap means, or null.
     *
     * [x] and [y] are **logical units** — view pixels divided by the
     * display density, the same space [setMetrics] is given — in panel
     * coordinates. A rotated panel is undone on the engine's side.
     */
    fun tapAction(x: Float, y: Float): String? =
        Native.tapAction(handle, x, y).ifEmpty { null }

    /** What a key means, or null. Takes an `KeyEvent.KEYCODE_*` value. */
    fun actionForKeyCode(keyCode: Int): String? =
        Native.actionForKeyCode(handle, keyCode).ifEmpty { null }

    /**
     * Apply what a tap or a key meant. You do not have to know which
     * action it was — hand back what you were given and read the outcome.
     */
    fun apply(action: String): ActionOutcome = when (Native.applyAction(handle, action)) {
        0 -> ActionOutcome.Changed
        1 -> ActionOutcome.Unchanged
        else -> ActionOutcome.Unhandled
    }

    // The text surface: what an accessibility tree, TTS, or a dictionary
    // popup consumes. Re-fetch after anything that redraws — a turn, a
    // reflow, a settings change.

    /**
     * The current page's text runs in reading order, or null until the
     * page is laid out. Empty for a page with nothing to speak (a comic).
     */
    fun pageTextRuns(): List<TextRun>? {
        val count = Native.pageTextRunCount(handle)
        if (count < 0) return null
        return (0 until count).mapNotNull { index ->
            val packed = Native.pageTextRunRange(handle, index)
            val rect = Native.pageTextRunRect(handle, index)
            if (packed < 0 || rect.size != 4) return@mapNotNull null
            TextRun(
                text = Native.pageTextRunText(handle, index),
                rect = android.graphics.RectF(
                    rect[0], rect[1], rect[0] + rect[2], rect[1] + rect[3]),
                locatorStart = (packed ushr 32).toInt(),
                locatorEnd = (packed and 0xffffffffL).toInt(),
            )
        }
    }

    /** The page as one string for a TTS utterance; `""` until laid out. */
    val speakableText: String get() = Native.speakableText(handle)

    /** The word table mapping TTS progress back to locator space. */
    fun pageWords(): List<WordSpan> {
        val flat = Native.pageWords(handle)
        return (flat.indices step 4).map { i ->
            WordSpan(flat[i], flat[i + 1], flat[i + 2], flat[i + 3])
        }
    }

    /**
     * The word under a panel point as a locator range, or null — off
     * text, on whitespace, on bare punctuation. Dictionary lookup's
     * question; feed the range to [rangeRects] or a selection.
     */
    fun wordAt(x: Float, y: Float): Pair<Int, Int>? {
        val packed = Native.wordAt(handle, x, y)
        if (packed < 0) return null
        return (packed ushr 32).toInt() to (packed and 0xffffffffL).toInt()
    }

    /** Page-space rects covering a locator range on the current page. */
    fun rangeRects(start: Int, end: Int): List<android.graphics.RectF> {
        val flat = Native.rangeRects(handle, start, end)
        return (flat.indices step 4).map { i ->
            android.graphics.RectF(
                flat[i], flat[i + 1], flat[i] + flat[i + 2], flat[i + 1] + flat[i + 3])
        }
    }

    /**
     * The library row this session's book was imported into, or null for
     * a book that never reached one — an OPDS stream, or a session opened
     * without a library directory.
     *
     * The join between the reading view and the shelf: opening a book is
     * what adds it, so this is how an app learns which [Book] it just
     * created and can put it in a collection, mark it, or find it again.
     */
    fun bookId(): Long? = Native.sessionBookId(handle).takeIf { it != 0L }

    override fun close() {
        if (handle != 0L) {
            Native.close(handle)
            handle = 0L
        }
    }
}

File: ios/Chapbook/Sources/Chapbook/Session.swift
import CChapbook
import Foundation

/// An open book.
///
/// Deliberately **not** `Sendable`. The handle underneath is `Send` and
/// not `Sync` — movable between threads, never shareable across them —
/// and that is exactly what a non-`Sendable` class expresses under
/// strict concurrency: region isolation can *transfer* a session into a
/// task for good, and using it from two isolation domains at once is the
/// data race Swift refuses to build. Marking this `@unchecked Sendable`
/// would be claiming the `Sync` the engine does not have; don't.
///
/// One consequence worth designing around rather than against: a
/// top-level `let` is a global in Swift's eyes and a global can never be
/// proven sent, so scope the session to a view controller or an actor,
/// not a singleton.
public final class Session {
    let raw: OpaquePointer
    private var waker: Unmanaged<WakerBox>?

    /// Open a book. The configuration is spent either way; on failure the
    /// error's `message` says why — install [`EngineLog`] first and the
    /// engine narrates its own failures too.
    public init(source: BookSource, configuration: SessionConfiguration) throws {
        let config = try configuration.makeRaw()
        // Every open consumes `config`, success or not.
        let session: OpaquePointer? =
            switch source {
            case .path(let url):
                cb_session_open_path(url.path, config)
            case .bytes(let data, let format):
                data.withUnsafeBytes { buffer in
                    cb_session_open_bytes(
                        buffer.bindMemory(to: UInt8.self).baseAddress,
                        buffer.count, format.rawValue, config)
                }
            case .fileDescriptor(let fd, let format):
                cb_session_open_fd(fd, format.rawValue, config)
            case .catalog(let url):
                cb_session_open_url(url.absoluteString, config)
            }
        guard let session else { throw ChapbookError.openFailure() }
        raw = session
    }

    deinit {
        cb_session_close(raw)
        waker?.release()
    }

    // MARK: Shape

    /// Reconfigure the page box. Call before the first render and on any
    /// size, scale or rotation change; the reader's place survives it.
    public func setMetrics(_ metrics: PageMetrics) throws {
        try check(cb_session_set_metrics(raw, metrics.raw))
    }

    public func title() -> String? {
        readString { cb_session_title(self.raw, $0, $1, $2) }
    }

    /// Pages in the current unit, at the current metrics.
    public func pageCount() throws -> Int {
        var count = 0
        try check(cb_session_page_count(raw, &count))
        return count
    }

    /// Units (chapters) in the book.
    public func spineLength() throws -> Int {
        var length = 0
        try check(cb_session_spine_len(raw, &length))
        return length
    }

    /// Where the reader is. Always the pair — a page number alone is
    /// meaningless across a unit boundary.
    public struct Position: Hashable, Sendable {
        public let spine: Int
        public let page: Int
    }

    public func position() throws -> Position {
        var raw = cb_position(spine: 0, page: 0)
        try check(cb_session_position(self.raw, &raw))
        return Position(spine: Int(raw.spine), page: Int(raw.page))
    }

    public func readingDirection() throws -> ReadingDirection {
        var raw: UInt32 = 0
        try check(cb_session_reading_direction(self.raw, &raw))
        return ReadingDirection(rawValue: raw) ?? .leftToRight
    }

    // MARK: Navigation

    /// Returns whether the position moved — use it, never a position
    /// comparison, to decide on a repaint.
    @discardableResult
    public func nextPage() throws -> Bool {
        var moved = false
        try check(cb_session_next_page(raw, &moved))
        return moved
    }

    @discardableResult
    public func prevPage() throws -> Bool {
        var moved = false
        try check(cb_session_prev_page(raw, &moved))
        return moved
    }

    // MARK: Input

    /// What a tap means, or `nil` for nothing. `point` is in the same
    /// logical panel coordinates the metrics were given — a `UITouch`
    /// location is already right, pass it as is. A rotated panel is
    /// undone on the engine's side.
    public func tapAction(at point: CGPoint) throws -> Action? {
        var raw: UInt32 = 0
        try check(cb_session_tap_action(self.raw, Float(point.x), Float(point.y), &raw))
        return raw == C.actionNone ? nil : Action(rawValue: raw)
    }

    /// Reconfigure the tap bands, as fractions of the page width, with
    /// `middle: nil` for a middle band that does nothing. The reading
    /// direction is never a parameter: it is the book's, so a tap cannot
    /// be resolved against the wrong edge. The default is thirds, middle
    /// bound to [`Action.toggleMenu`].
    public func setTapZones(
        prevFraction: CGFloat, nextFraction: CGFloat, middle: Action?
    ) throws {
        try check(
            cb_session_set_tap_zones(
                raw, Float(prevFraction), Float(nextFraction),
                middle?.rawValue ?? C.actionNone))
    }

    /// Apply a reader intent and learn both what happened and who owns
    /// the event — see [`ActionOutcome`].
    public func apply(_ action: Action) throws -> ActionOutcome {
        var raw: UInt32 = 0
        try check(cb_session_apply(self.raw, action.rawValue, &raw))
        return ActionOutcome(rawValue: raw) ?? .unhandled
    }

    // MARK: Settings

    public func settings() throws -> ReadingSettings {
        var raw = cb_settings(
            base_font_px: 0, line_height: 0, justify: false,
            publisher_styles: false, theme: 0)
        try check(cb_session_settings(self.raw, &raw))
        return ReadingSettings(raw: raw)
    }

    /// Apply settings, keeping the reader's place across the reflow.
    public func setSettings(_ settings: ReadingSettings, scope: SettingsScope) throws {
        try check(cb_session_set_settings(raw, settings.raw, scope.rawValue))
    }

    // MARK: Lifecycle and memory

    /// Save the position and let go of everything reconstructible,
    /// including the database and its POSIX locks — call it from the last
    /// callback the platform guarantees (`willResignActive` /
    /// `didEnterBackground`). The session stays usable afterwards.
    public func suspend() throws {
        try check(cb_session_suspend(raw))
    }

    /// Drop cached layouts and decoded pages, for a memory warning. The
    /// current page rebuilds on the next render.
    public func releaseCaches() throws {
        try check(cb_session_release_caches(raw))
    }

    public func cacheBytes() throws -> Int {
        var bytes = 0
        try check(cb_session_cache_bytes(raw, &bytes))
        return bytes
    }

    // MARK: Fonts, diagnosed

    /// How many faces the font source produced — worth logging once at
    /// startup, since font failures are the quiet kind.
    public func fontFaceCount() throws -> Int {
        var count = 0
        try check(cb_session_font_face_count(raw, &count))
        return count
    }

    /// Generic families that resolved to a name no loaded face carries,
    /// one `generic=family` per line; `nil` when everything resolved.
    public func unresolvedFontGenerics() -> String? {
        let report = readString { cb_session_font_unresolved(self.raw, $0, $1, $2) }
        return report.flatMap { $0.isEmpty ? nil : $0 }
    }

    // MARK: Background loads

    /// Comic and PDF pages decode on the engine's loader thread. The
    /// handler fires **on that thread** when something new is ready — do
    /// nothing there but hop to the main actor and call
    /// [`pollLoaded()`], repainting if it answers `true`.
    public func onWake(_ handler: @escaping @Sendable () -> Void) throws {
        let box = Unmanaged.passRetained(WakerBox(handler))
        // A Swift closure with captures is not a C function pointer,
        // which is exactly what `user` exists for.
        try check(
            cb_session_set_waker(
                raw,
                { user in
                    guard let user else { return }
                    Unmanaged<WakerBox>.fromOpaque(user).takeUnretainedValue().handler()
                },
                box.toOpaque()))
        waker?.release()
        waker = box
    }

    /// Take delivery of anything the loader finished; `true` means the
    /// visible page changed and a repaint is worth doing.
    public func pollLoaded() throws -> Bool {
        var changed = false
        try check(cb_session_poll_loaded(raw, &changed))
        return changed
    }

    public func hasPendingLoads() throws -> Bool {
        var pending = false
        try check(cb_session_has_pending_loads(raw, &pending))
        return pending
    }
}

final class WakerBox: @unchecked Sendable {
    let handler: @Sendable () -> Void
    init(_ handler: @escaping @Sendable () -> Void) { self.handler = handler }
}
